src/hangar/typesystem/descriptors.py

```diff
@@ -0,0 +1,187 @@
+            raise TypeError(f'not all elements are {expected_element_types} type(s) in {value}')
+        super_set(self, instance, value)
+    cls.__set__ = __set__
+    return cls
+
+
+def OneOf(expected_values, cls=None):
+    # Decorator for enforcing values
+    if cls is None:
+        return lambda cls: OneOf(expected_values, cls)
+
+    super_set = cls.__set__
+    def __set__(self, instance, value):
+        if value not in expected_values:
+            raise ValueError(f'expected one of {expected_values} recieved {value}')
+        super_set(self, instance, value)
+    cls.__set__ = __set__
+    return cls
+
+
+def MaxSized(cls):
+    # Decorator for allowing sized values
+    super_init = cls.__init__
+    def __init__(self, name=None, **opts):
+        if 'size' not in opts:
+            raise TypeError('missing size option')
+        self.size = opts['size']
+        super_init(self, name, **opts)
+    cls.__init__ = __init__
+    super_set = cls.__set__
+    def __set__(self, instance, value):
+        if len(value) > self.size:
+            raise ValueError('size must be < ' + str(self.size))
+        super_set(self, instance, value)
+    cls.__set__ = __set__
+    return cls
+
+
+def DictItems(expected_keys_required, expected_values, cls=None):
+    # check a dictionary for the existence of keys. expected_keys should be a dictionary of keys,
+    # with bool values set to indicate if they are required or not. expected_values should be
+    # mapping of same keys to list of acceptable values.
+    if cls is None:
+        return lambda cls: DictItems(expected_keys_required, expected_values, cls)
+
+    super_set = cls.__set__
+    def __set__(self, instance, value):
+        if not isinstance(value, dict):
+            raise TypeError(f'expected {dict}, recieved {type(value)}')
+        for expected_key, required in expected_keys_required.items():
+            try:
+                if value[expected_key] not in expected_values[expected_key]:
+                    raise ValueError(f'{value[expected_key]} invalid for key {expected_key}')
+            except KeyError as e:
+                if required:
+                    raise e
+        for recieved_key in value.keys():
+            if recieved_key not in expected_keys_required:
+                raise TypeError(f'Not supposed to have key {recieved_key}')
+        super_set(self, instance, value)
+    cls.__set__ = __set__
+    return cls
+
+
+@Typed(str)
+class String(Descriptor):
+    pass
+
+
+@DictItems(expected_keys_required={},
+           expected_values={},)
+class EmptyDict(Descriptor):
+    pass
+
+
+@Typed((dict, type(None)))
+class OptionalDict(Descriptor):
+    pass
+
+
+@Typed((str, type(None)))
+class OptionalString(Descriptor):
+    pass
+
+
+@Typed(tuple)
+class Tuple(Descriptor):
+    pass
+
+
+@MaxSized
+@TypedSequence(int)
+class SizedIntegerTuple(Tuple):
+    pass
+
+
+class checkedmeta(type):
```

Comment:
`__set_name__` does exactly this. We don't need this metaclass perhaps.

Reply:
I'll take a look, thanks for pointing that out!